### PR TITLE
mvcc: don't write below provisional commit timestamp on WriteTooOld errors

### LINKED ISSUE #34853: roachtest: cdc/tpcc-1000/rangefeed=true failed
SHA: https://github.com/cockroachdb/cockroach/commits/10f8010fa5778e740c057905e2d7664b5fd5d647

Parameters:

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stress instead of stressrace and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
stdbuf -oL -eL \
make stressrace TESTS=cdc/tpcc-1000/rangefeed=true PKG=roachtest TESTTIMEOUT=5m STRESSFLAGS='-maxtime 20m -timeout 10m' 2>&1 | tee /tmp/stress.log
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=1135549&tab=buildLog

```
The test failed on master:
	cluster.go:1226,cdc.go:612,cdc.go:124,cluster.go:1564,errgroup.go:57: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/bin/roachprod run teamcity-1135549-cdc-tpcc-1000-rangefeed-true:4 -- ./workload run tpcc --warehouses=1000 --duration=120m --tolerate-errors {pgurl:1-3}  returned:
		stderr:
		
		stdout:
		0.0      0.0 newOrder
		     38s      688            0.0           22.4      0.0      0.0      0.0      0.0 orderStatus
		     38s      688            0.0          197.1      0.0      0.0      0.0      0.0 payment
		     38s      688            0.0           20.4      0.0      0.0      0.0      0.0 stockLevel
		E190213 08:12:19.560089 1 workload/cli/run.go:402  error in payment: dial tcp 10.142.0.98:26257: connect: connection refused
		     39s      909            0.0           19.5      0.0      0.0      0.0      0.0 delivery
		     39s      909            0.0          138.4      0.0      0.0      0.0      0.0 newOrder
		     39s      909            0.0           21.9      0.0      0.0      0.0      0.0 orderStatus
		     39s      909            0.0          192.1      0.0      0.0      0.0      0.0 payment
		     39s      909            0.0           19.9      0.0      0.0      0.0      0.0 stockLevel
		E190213 08:12:20.560203 1 workload/cli/run.go:402  error in payment: dial tcp 10.142.0.98:26257: connect: connection refused
		: signal: killed
	cluster.go:1585,cdc.go:212,cdc.go:367,test.go:1212: unexpected node event: 3: dead

```

### PR DESCRIPTION
**Pull request by nvb**
Fixes #34853.

This change fixes #34853, which appears to have been broken by
https://github.com/cockroachdb/cockroach/commit/57d02014e661aa4f153e56c558950d4c55c7877d#diff-41e9e1b7829f8e25daeb564f42c26017L1514.
That change made it so that intents are now written at a transaction's
provisional commit timestamp instead of its original timestamp (which is
fantastic!). In doing so, it removed special-purpose code to forward the
`MVCCLogicalOpDetails` timestamp to the provisional commit timestamp of the
transaction because it assumed that the new `writeTimestamp` variable that the
change introduced already assumed this role.

Unfortunately, the change allowed this `writeTimestamp` variable to regress
below the transaction's provisional commit timestamp in subtle cases where a
committed value ended up between the transaction's original timestamp and its
provisional commit timestamp (a case that generates `WriteTooOldError`s). In
this case, we were replacing the `writeTimestamp` with the committed value's
timestamp + 1 logical tick. This regression allowed `MVCCWriteIntentOp`s to look
like they were ignoring closed timestamps.

The fix was to either introduce the special purpose code again or to avoid
letting `writeTimestamp` regress in this `WriteTooOldError` code path. This PR
chose the latter solution.

In doing so, this PR also improves upon the benefit already provided by #32487 -
it reduces the number of intents that need to be moved on transaction commit.
This is because it ensures that if a transaction has an original (read)
timestamp below a committed value and a provisional commit (write) timestamp
above a committed value, an intent left by that transaction will be in a
"committable" state without a need to rewrite it during intent resolution.

Release note: None

### CODE REVIEW

**Review by tbg — COMMENTED** [review]
Great that you got to the bottom of this so fast! Man, that method is hard to follow. I gave this what I hope is a thorough review but will give it another one when you've figured out the lost update.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/35086#-:-LZ9CsulDMyGYXjw7OWe:bwz2pit)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @danhhz, @nvanbenschoten, and @tbg)

___
*[pkg/storage/engine/mvcc.go, line 1466 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/35086#-LZ9AS5U7Y7rjuXMEW1D:-LZ9AS5VBPA-3-bw6GBt:b-92flba) ([raw file](https://github.com/cockroachdb/cockroach/blob/ab8ccbf7f7cdc95156a604ec301c4f83c02ed1a9/pkg/storage/engine/mvcc.go#L1466)):*
> ```Go
> 			}
> 		} else if !metaTimestamp.Less(readTimestamp) {
> 			// This is the case where we're trying to write under a committed
> ```

Not necessarily, right? So far we only know that there's a committed write above the read timestamp (but the write timestamp could be above the committed value). Nvm, you mull on that below.
___
*[pkg/storage/engine/mvcc.go, line 1483 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/35086#-LZ9Amlf05WwcboXJSnN:-LZ9Amlf05WwcboXJSnO:bto9ndj) ([raw file](https://github.com/cockroachdb/cockroach/blob/ab8ccbf7f7cdc95156a604ec301c4f83c02ed1a9/pkg/storage/engine/mvcc.go#L1483)):*
> ```Go
> 			// we'll need to refresh anyway, right? It might also be interacting
> 			// poorly with our logic in `evaluateBatch`. Any idea @bdarnell or
> 			// @tbg?
> ```

For snapshot transactions, this is just lost update -- two txn start at origts=100, their commit ts advances to 150 and 200 respectively, and then txn1 writes key=value1 (and commits) and txn2 cputs key from nil to value2, in effect wiping out value1 without ever having to observe it. I also don't see how this applies to serializable txns, though. As you mention, if we lay down an intent we're going to have to refresh anyway, and that should see the update that would be lost and force a restart. Worth getting to the bottom of even though I'm not sure avoiding the WTO is a good idea here. After all, what are the odds the write will survive the refresh? At that point we might as well check right here whether it would survive a refresh, though that seems to be optimizing for something we don't know is worth the added complexity.

PS this method probably has a cyclomatic complexity of MaxUint64... not that I see any easy way of changing that.
___
*[pkg/storage/engine/mvcc.go, line 1484 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/35086#-LZ9CQo3BJMJ-YKJIoh5:-LZ9CQo3BJMJ-YKJIoh6:b-d2ncv) ([raw file](https://github.com/cockroachdb/cockroach/blob/ab8ccbf7f7cdc95156a604ec301c4f83c02ed1a9/pkg/storage/engine/mvcc.go#L1484)):*
> ```Go
> 			// poorly with our logic in `evaluateBatch`. Any idea @bdarnell or
> 			// @tbg?
> 			writeTimestamp.Forward(metaTimestamp.Next())
> ```

Does this writeTimestamp ever make it back to the txn in the response?
___
*[pkg/storage/engine/mvcc.go, line 1497 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/35086#-LZ9Bx2kECc-INAylwUT:-LZ9Bx2l9HiqZT3JfCAL:b-topte8) ([raw file](https://github.com/cockroachdb/cockroach/blob/ab8ccbf7f7cdc95156a604ec301c4f83c02ed1a9/pkg/storage/engine/mvcc.go#L1497)):*
> ```Go
> 			} else {
> 				// Outside of a transaction, read the latest value and advance
> 				// the write timestamp to the latest value's timestamp + 1. The
> ```

This part of the comment seems out of place because this was already done (also before this PR).
___
*[pkg/storage/engine/mvcc.go, line 1498 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/35086#-LZ9Dg6DEAqurp5HFTmn:-LZ9Dg6DEAqurp5HFTmo:b-10tynw) ([raw file](https://github.com/cockroachdb/cockroach/blob/ab8ccbf7f7cdc95156a604ec301c4f83c02ed1a9/pkg/storage/engine/mvcc.go#L1498)):*
> ```Go
> 				// Outside of a transaction, read the latest value and advance
> 				// the write timestamp to the latest value's timestamp + 1. The
> 				// new timestamp is returned to the caller in maybeTooOldErr.
> ```

Confused by this. So we do actually write but then also return maybeTooOldErr? I doubt that.

**Review by nvb — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/35086)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @danhhz and @tbg)

___
*[pkg/storage/engine/mvcc.go, line 1466 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/35086#-LZ9AS5U7Y7rjuXMEW1D:-LZC6Msm0N37IPw2Z9da:bb6sh2i) ([raw file](https://github.com/cockroachdb/cockroach/blob/ab8ccbf7f7cdc95156a604ec301c4f83c02ed1a9/pkg/storage/engine/mvcc.go#L1466)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Not necessarily, right? So far we only know that there's a committed write above the read timestamp (but the write timestamp could be above the committed value). Nvm, you mull on that below.
</blockquote></details>

Ack.
___
*[pkg/storage/engine/mvcc.go, line 1483 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/35086#-LZ9Amlf05WwcboXJSnN:-LZCH_8D4Gg-wXcCNocM:b-6r6siy) ([raw file](https://github.com/cockroachdb/cockroach/blob/ab8ccbf7f7cdc95156a604ec301c4f83c02ed1a9/pkg/storage/engine/mvcc.go#L1483)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

For snapshot transactions, this is just lost update -- two txn start at origts=100, their commit ts advances to 150 and 200 respectively, and then txn1 writes key=value1 (and commits) and txn2 cputs key from nil to value2, in effect wiping out value1 without ever having to observe it. I also don't see how this applies to serializable txns, though. As you mention, if we lay down an intent we're going to have to refresh anyway, and that should see the update that would be lost and force a restart. Worth getting to the bottom of even though I'm not sure avoiding the WTO is a good idea here. After all, what are the odds the write will survive the refresh? At that point we might as well check right here whether it would survive a refresh, though that seems to be optimizing for something we don't know is worth the added complexity.

PS this method probably has a cyclomatic complexity of MaxUint64... not that I see any easy way of changing that.
</blockquote></details>

Mystery solved!
___
*[pkg/storage/engine/mvcc.go, line 1484 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/35086#-LZ9CQo3BJMJ-YKJIoh5:-LZC6gxp1hxj8yi5S1Jg:b-g45egp) ([raw file](https://github.com/cockroachdb/cockroach/blob/ab8ccbf7f7cdc95156a604ec301c4f83c02ed1a9/pkg/storage/engine/mvcc.go#L1484)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Does this writeTimestamp ever make it back to the txn in the response?
</blockquote></details>

Yes, the `writeTimestamp` makes it back into the txn response here: https://github.com/cockroachdb/cockroach/blob/df358aca2850a5f6300e0da46f30ba3ae7d13bf4/pkg/storage/replica_evaluate.go#L288
___
*[pkg/storage/engine/mvcc.go, line 1497 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/35086#-LZ9Bx2kECc-INAylwUT:-LZC7FWBAYGZ1SlPr1Tr:b-bw8e14) ([raw file](https://github.com/cockroachdb/cockroach/blob/ab8ccbf7f7cdc95156a604ec301c4f83c02ed1a9/pkg/storage/engine/mvcc.go#L1497)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

This part of the comment seems out of place because this was already done (also before this PR).
</blockquote></details>

Updated.
___
*[pkg/storage/engine/mvcc.go, line 1498 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/35086#-LZ9Dg6DEAqurp5HFTmn:-LZC7GGG7_1CD957Coot:b-mksb09) ([raw file](https://github.com/cockroachdb/cockroach/blob/ab8ccbf7f7cdc95156a604ec301c4f83c02ed1a9/pkg/storage/engine/mvcc.go#L1498)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Confused by this. So we do actually write but then also return maybeTooOldErr? I doubt that.
</blockquote></details>

It's surprising, but yes we do. Dating all the way back to https://github.com/cockroachdb/cockroach/commit/4220d4ab056dc226e3740b1b934161ecfbb24222!

**Review by tbg — COMMENTED** [review]
> Reviewed 1 of 2 files at r1, 1 of 1 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/35086)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @danhhz and @nvanbenschoten)

___
*[pkg/storage/engine/mvcc.go, line 1483 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/35086#-LZ9Amlf05WwcboXJSnN:-LZE1QF27q3vRTZnBaQh:b-mcfn37) ([raw file](https://github.com/cockroachdb/cockroach/blob/ab8ccbf7f7cdc95156a604ec301c4f83c02ed1a9/pkg/storage/engine/mvcc.go#L1483)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Mystery solved!
</blockquote></details>

How does this mechanism work exactly? If a CPut runs into a WTOErr, it "locally" (store?) retries the batch at the suggested timestamp and then returns the txn retry error? How does your attempted change break that? Shouldn't we be removing the WTO, but still returning a txn that needs to refresh if it wants to have any chance of committing? Why does it get to commit anyway?
___
*[pkg/storage/engine/mvcc.go, line 1498 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/35086#-LZ9Dg6DEAqurp5HFTmn:-LZE2Ipi4FTJ_fggw99r:bh6t80h) ([raw file](https://github.com/cockroachdb/cockroach/blob/ab8ccbf7f7cdc95156a604ec301c4f83c02ed1a9/pkg/storage/engine/mvcc.go#L1498)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

It's surprising, but yes we do. Dating all the way back to https://github.com/cockroachdb/cockroach/commit/4220d4ab056dc226e3740b1b934161ecfbb24222!
</blockquote></details>

But we're not in a transaction in this branch? The write must not be committed, so why do it?
This is probably not the right PR to be pulling on this thread, but add a TODO or some explanation, if applicable.

**Review by nvb — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/35086)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @danhhz and @tbg)

___
*[pkg/storage/engine/mvcc.go, line 1483 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/35086#-LZ9Amlf05WwcboXJSnN:-LZF_b4FAkb6ZnmNb44X:bkes96m) ([raw file](https://github.com/cockroachdb/cockroach/blob/ab8ccbf7f7cdc95156a604ec301c4f83c02ed1a9/pkg/storage/engine/mvcc.go#L1483)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

How does this mechanism work exactly? If a CPut runs into a WTOErr, it "locally" (store?) retries the batch at the suggested timestamp and then returns the txn retry error? How does your attempted change break that? Shouldn't we be removing the WTO, but still returning a txn that needs to refresh if it wants to have any chance of committing? Why does it get to commit anyway?
</blockquote></details>

The mechanism is all in [`evaluateBatch`](https://github.com/cockroachdb/cockroach/blob/251bc8eb5bfbe4668b2dc26abd8eb3e77336da33/pkg/storage/replica_evaluate.go#L240). The idea isn't to locally retry anything. Instead, that function's purpose is to run all commands in a batch even if they are throwing WTO errors to determine the highest timestamp needed for later retries.

The function then makes a few decision. First, for non-transactional requests, it always returns the largest WriteTooOldError. So to answer the question below, we will still return the error instead of committing, but we will wait until the entire non-transactional batch has finished evaluating so that we can return the "best" error.

For transactional requests, things are more interesting. For suitable request types like blind Puts, it sets the batch responses timestamp and its WriteTooOld flag. It can then go ahead and writes these intents. However, for others that are not blind like CPut, it returns the WriteTooOld error immediately instead of writing the intents. This allows these request types to avoid ever needing to be refreshed. This is also what broke when I changed this to rely on CPuts refreshing.

I added some more commentary in the code.
___
*[pkg/storage/engine/mvcc.go, line 1498 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/35086#-LZ9Dg6DEAqurp5HFTmn:-LZFbq4m1glxJ8XNDnjH:b-bang1d) ([raw file](https://github.com/cockroachdb/cockroach/blob/ab8ccbf7f7cdc95156a604ec301c4f83c02ed1a9/pkg/storage/engine/mvcc.go#L1498)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

But we're not in a transaction in this branch? The write must not be committed, so why do it?
This is probably not the right PR to be pulling on this thread, but add a TODO or some explanation, if applicable.
</blockquote></details>

Answered above.

**Review by tbg — APPROVED** [review]
> Reviewed 1 of 1 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/35086)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @danhhz)

**Comment by nvanbenschoten** [comment]
bors r=tbg
